File: touchfish-core/src/domain/entity/fish.rs
use serde::{Deserialize, Serialize};
use serde_with::skip_serializing_none;
use strum_macros::{Display, EnumString};
use yfunc_rust::{prelude::*, YBytes, YTime};

use crate::FishPreview;

#[yfunc]
#[derive(Serialize, Debug)]
pub struct Fish {
    pub id: i32,
    pub identity: String,
    pub count: i32,
    pub fish_type: FishType,
    pub fish_data: YBytes,
    pub data_info: DataInfo,
    pub desc: String,
    pub tags: Vec<String>,
    pub is_marked: bool,
    pub is_locked: bool,
    pub extra_info: String,
    pub create_time: YTime,
    pub update_time: YTime,
}

impl Fish {

    pub fn to_preview_json(&self) -> YRes<String> {
        FishPreview::from_fish(self).trace(
            ctx!("serialize Fish to preview style json string -> build FishPreview from Fish: FishPreview::from_fish failed", self.identity)
        )?.to_pretty_json_str().trace(
            ctx!("serialize Fish to preview style json string -> serialize FishPreview to json: FishPreview::to_pretty_json_str failed", self.identity)
        )
    }

}

#[yfunc]
#[derive(Serialize, Deserialize, Debug, EnumString, Display, PartialEq, Eq, Hash, Clone, Copy)]
pub enum FishType {
    Text,
    Image,
    Other,
}

#[yfunc]
#[skip_serializing_none]
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct DataInfo {
    pub byte_count: Option<usize>,
    // Text
    pub char_count: Option<usize>,
    pub word_count: Option<usize>,
    pub row_count: Option<usize>,
    // Image
    pub width: Option<usize>,
    pub height: Option<usize>,
}

impl DataInfo {

    pub fn new() -> DataInfo {
        DataInfo { 
            byte_count: None, 
            char_count: None, word_count: None, row_count: None,
            width: None, height: None,
        }
    }

}

File: touchfish-python-sdk/src/touchfish/fish.py
from enum import Enum
import base64
from yfunc import *

class FishType(Enum):
    Text = 'Text'
    Image = 'Image'
    Other = 'Other'

class DataInfo:
    byte_count: int | None
    char_count: int | None
    word_count: int | None
    row_count: int | None
    width: int | None
    height: int | None

    def __init__(self, arg: dict):
        self.byte_count = arg.get('byte_count', None)
        self.char_count = arg.get('char_count', None)
        self.word_count = arg.get('word_count', None)
        self.row_count = arg.get('row_count', None)
        self.width = arg.get('width', None)
        self.height = arg.get('height', None)
    
    def __str__(self) -> str:
        return str(self.__dict__)
    
    def __repr__(self) -> str:
        return str(self.__dict__)
        
class Fish:
    uid: str
    identity: str
    count: int
    fish_type: FishType
    fish_data: bytes
    data_info: DataInfo
    desc: str
    tags: list[str]
    is_marked: bool
    is_locked: bool
    extra_info: dict
    create_time: str
    update_time: str

    def __init__(self, arg: dict):
        self.uid = arg['uid']
        self.identity = arg['identity']
        self.count = arg['count']
        self.fish_type = FishType(arg['fish_type'])
        self.fish_data = base64.b64decode(arg['fish_data'])
        self.data_info = DataInfo(arg['data_info'])
        self.desc = arg['desc']
        self.tags = arg['tags']
        self.is_marked = arg['is_marked']
        self.is_locked = arg['is_locked']
        self.extra_info = ystr(arg['extra_info']).json().to_dic()
        self.create_time = arg['create_time']
        self.update_time = arg['update_time']

    def text_data(self) -> str:
        return self.fish_data.decode('utf-8')

    def __str__(self) -> str:
        return str(self.__dict__)
    
    def __repr__(self) -> str:
        return str(self.__dict__)
